File: ARRAY/Searching/binarySearch.cpp
#include <iostream>
using namespace std;

int binarySearch(int arr[], int size, int target){
    int start = 0;
    int end = size - 1;

    while(start <= end){
        int mid = start + (end - start) / 2;

        if(target > arr[mid]){
            start = mid + 1;
        }
        else if(target < arr[mid]){
            end = mid - 1;
        }
        else{
            return mid;
        }
    }
    return -1;
}

int main(){
    int size;
    cout << "Enter size of array : ";
    cin >> size;

    int arr[size];
    cout << "Enter element of array : ";
    for(int i=0;i<size;i++){
        cin >> arr[i];
    }

    cout << "You entered array is : ";
    for(int i=0;i<size;i++){
        cout << arr[i] << " ";
    }

    int target;
    cout << "\nEnter element to be search : ";
    cin >> target;

    cout << "*********************************************\n";
    int result = binarySearch(arr,size,target);
    if(result == -1){
        cout << target << " element is not present in the array";
    }
    else{
        cout << target << " element is present in the array at index " << result;
    }

    return 0;
}

File: BinarySearch/BinarySearch.java
package BinarySearch;
import java.util.*;
public class BinarySearch {
    
    // Binary Search Function
    public static int binarySearch(int[] arr, int target) {
        int left = 0;
        int right = arr.length - 1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;  // To avoid overflow
            
            // Check if target is present at mid
            if (arr[mid] == target) {
                return mid;
            }
            
            // If target is greater, ignore the left half
            if (arr[mid] < target) {
                left = mid + 1;
            }
            // If target is smaller, ignore the right half
            else {
                right = mid - 1;
            }
        }
        
        // If we reach here, the element is not present
        return -1;
    }
    
    public static void main(String[] args) {
        int[] arr = {2, 3, 4, 10, 40};
        int target = 10;
        
        // The array must be sorted for binary search
        Arrays.sort(arr);
        
        int result = binarySearch(arr, target);
        
        if (result == -1) {
            System.out.println("Element not present in array");
        } else {
            System.out.println("Element found at index " + result);
        }
    }
}
